Frameworks/CoreGraphics/DWriteFontCollectionHelper.mm

```diff
@@ -0,0 +1,207 @@
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import "DWriteFontCollectionHelper.h"
+
+#import <vector>
+
+#import <StringHelpers.h>
+
+using namespace Microsoft::WRL;
+
+static const wchar_t* TAG = L"_DWriteFontCollectionHelper";
+
+CFMutableArrayRef DWriteFontCollectionHelper::CopyFontFamilyNames() {
+    ComPtr<IDWriteFontCollection> fontCollection = _GetFontCollection();
+    if (!fontCollection) {
+        // Return empty array for an empty font collection
+        return CFArrayCreateMutable(kCFAllocatorSystemDefault, 0, &kCFTypeArrayCallBacks);
+    }
+
+    // Get the number of font families in the collection.
+    uint32_t count = fontCollection->GetFontFamilyCount();
+    CFMutableArrayRef ret = CFArrayCreateMutable(kCFAllocatorSystemDefault, count, &kCFTypeArrayCallBacks);
+    for (uint32_t i = 0; i < count; ++i) {
+        // Get the font family.
+        ComPtr<IDWriteFontFamily> fontFamily;
+        RETURN_NULL_IF_FAILED(fontCollection->GetFontFamily(i, &fontFamily));
+
+        // Get a list of localized strings for the family name.
+        ComPtr<IDWriteLocalizedStrings> familyNames;
+        RETURN_NULL_IF_FAILED(fontFamily->GetFamilyNames(&familyNames));
+
+        CFStringRef name = _CFStringFromLocalizedString(familyNames.Get());
+        if (CFStringGetLength(name) == 0) {
+            TraceError(TAG, L"Failed to convert the localized string to wide string.");
+            continue;
+        }
+
+        CFArrayAppendValue(ret, name);
+    }
+
+    return ret;
+}
+
+CFMutableArrayRef DWriteFontCollectionHelper::CopyFontNamesForFamilyName(CFStringRef familyName) {
+    CFMutableArrayRef ret = CFArrayCreateMutable(kCFAllocatorSystemDefault, 0, &kCFTypeArrayCallBacks);
+
+    ComPtr<IDWriteFontList> fontList;
+    if (FAILED(_GetFontListForFamilyName(familyName, &fontList))) {
+        return ret;
+    }
+
+    size_t fontCount = fontList->GetFontCount();
+    for (size_t i = 0; i < fontCount; i++) {
+        ComPtr<IDWriteFont> font;
+        if (FAILED(fontList->GetFont(i, &font))) {
+            continue;
+        }
+
+        ComPtr<IDWriteLocalizedStrings> fullName;
+        BOOL exist = FALSE;
+        if (FAILED(font->GetInformationalStrings(DWRITE_INFORMATIONAL_STRING_FULL_NAME, &fullName, &exist))) {
+            continue;
+        }
+
+        if (exist) {
+            CFStringRef name = _CFStringFromLocalizedString(fullName.Get());
+            if (CFStringGetLength(name) == 0) {
+                TraceError(TAG, L"Failed to convert the localized string to wide string.");
+                continue;
+            }
+
+            CFArrayAppendValue(ret, name);
+        }
+    }
+
+    return ret;
+}
+
+std::shared_ptr<_DWriteFontProperties> DWriteFontCollectionHelper::GetFontPropertiesFromUppercaseFontName(
+    const woc::unique_cf<CFStringRef>& upperFontName) {
+    if (!m_propertiesMap) {
```

Comment:
was thinking about this over the weekend and probably yes

---
In reply to: [90909514](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90909514) [](ancestors = 90909514)